Request to download Octo client does not support a proxy

The `RestClient` used to contact https://g.octopushq.com/LatestTools does not have a way to be configured with a proxy.

Hi Matthew

We tested with the below version but it still tries to connect via proxy, same error as before

2018-10-28T22:40:25.1934210Z ##[section]Starting: Create Octopus Release: Endeavour Demo
2018-10-28T22:40:25.1934210Z ==============================================================================
2018-10-28T22:40:25.1934210Z Task         : Create Octopus Release
2018-10-28T22:40:25.1934210Z Description  : Create a Release in Octopus Deploy
2018-10-28T22:40:25.1934210Z Version      : 3.0.184
2018-10-28T22:40:25.1934210Z Author       : Octopus Deploy
2018-10-28T22:40:25.1934210Z Help         : Version: 3.0.184. [More Information](https://g.octopushq.com/TFS-VSTS)
2018-10-28T22:40:25.1934210Z ==============================================================================
2018-10-28T22:40:26.1934657Z SystemVssConnection exists true
2018-10-28T22:40:26.2090873Z 592dba07-b077-45ff-8728-c39ea74be1e6 exists true
2018-10-28T22:40:38.2720995Z ##[error]Error: getaddrinfo EAI_AGAIN g.octopushq.com:443
2018-10-28T22:40:38.2720995Z ##[error]Failed to deploy release getaddrinfo EAI_AGAIN g.octopushq.com:443
2018-10-28T22:40:38.2720995Z ##[section]Finishing: Create Octopus Release: Endeavour Demo


Thanks
Praveen


From: Matthew Casperson <notifications@github.com>
Sent: 29 October 2018 12:25 PM
To: OctopusDeploy/OctoTFS <OctoTFS@noreply.github.com>
Cc: Praveen Raniga <Praveen.Raniga@asb.co.nz>; Mention <mention@noreply.github.com>
Subject: Re: [OctopusDeploy/OctoTFS] Request to download Octo client does not support a proxy (#116)


Hi @praniga<https://github.com/praniga> , version 3.0.184 include the octo client for restricted environments.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/OctopusDeploy/OctoTFS/issues/116#issuecomment-433752343>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AOr2LlvM9eaEcm9nBmehYW4mZtTJBMn2ks5upjzXgaJpZM4Xp9gu>.

This email may contain information which is confidential and/or subject to legal privilege.  If you are not the intended recipient, please immediately notify the sender and delete the email.

--- a/source/tasks/BuildInformation/index.ts
+++ b/source/tasks/BuildInformation/index.ts
@@ -3,7 +3,7 @@ import { ToolRunner } from "azure-pipelines-task-lib/toolrunner";
 import * as utils from "../Utils";
 import * as path from "path";
 
-import { connectionArguments, includeArguments, argument, argumentEnquote, argumentIfSet, getOverwriteModeFromReplaceInput, multiArgument } from "../Utils";
+import { connectionArguments, includeAdditionalArgumentsAndProxyConfig, argument, argumentEnquote, argumentIfSet, getOverwriteModeFromReplaceInput, multiArgument } from "../Utils";
 
 export interface IOctopusBuildInformation {
     BuildEnvironment: string;
@@ -66,7 +66,7 @@ async function run() {
             argument("version", packageVersion),
             argumentEnquote("file", buildInformationFile),
             argument("overwrite-mode", overwriteMode),
-            includeArguments(additionalArguments),
+            includeAdditionalArgumentsAndProxyConfig(connection.url, additionalArguments),
         ];
 
         const code: Number = await octo

--- a/source/tasks/CreateOctopusRelease/CreateOctopusReleaseV3/index.ts
+++ b/source/tasks/CreateOctopusRelease/CreateOctopusReleaseV3/index.ts
@@ -1,6 +1,6 @@
 import * as tasks from "azure-pipelines-task-lib/task";
 import * as utils from "../../Utils";
-import { multiArgument, connectionArguments, includeArguments, flag, argumentEnquote, argumentIfSet } from "../../Utils/tool";
+import { multiArgument, connectionArguments, includeAdditionalArgumentsAndProxyConfig, flag, argumentEnquote, argumentIfSet } from "../../Utils/tool";
 
 async function run() {
     try {
@@ -28,7 +28,7 @@ async function run() {
             linkedReleaseNotes = await utils.getLinkedReleaseNotes(vstsConnection, changesetCommentReleaseNotes, workItemReleaseNotes);
         }
 
-        const realseNotesFile = utils.createReleaseNotesFile(() => {
+        const releaseNotesFile = utils.createReleaseNotesFile(() => {
             return utils.generateReleaseNotesContent(environmentVariables, linkedReleaseNotes, customReleaseNotes);
         }, environmentVariables.defaultWorkingDirectory);
 
@@ -43,8 +43,8 @@ async function run() {
             flag("progress", deployToEnvironments.length > 0 && deploymentProgress),
             multiArgument(argumentEnquote, "tenant", deployForTenants),
             multiArgument(argumentEnquote, "tenanttag", deployForTenantTags),
-            argumentEnquote("releaseNotesFile", realseNotesFile),
-            includeArguments(additionalArguments),
+            argumentEnquote("releaseNotesFile", releaseNotesFile),
+            includeAdditionalArgumentsAndProxyConfig(octoConnection.url, additionalArguments),
         ];
 
         const code: Number = await octo

--- a/source/tasks/CreateOctopusRelease/CreateOctopusReleaseV4/index.ts
+++ b/source/tasks/CreateOctopusRelease/CreateOctopusReleaseV4/index.ts
@@ -1,6 +1,6 @@
 import * as tasks from "azure-pipelines-task-lib/task";
 import * as utils from "../../Utils";
-import { multiArgument, connectionArguments, includeArguments, flag, argumentEnquote, argumentIfSet } from "../../Utils";
+import { multiArgument, connectionArguments, includeAdditionalArgumentsAndProxyConfig, flag, argumentEnquote, argumentIfSet } from "../../Utils";
 
 async function run() {
     try {
@@ -51,7 +51,7 @@ async function run() {
             configure.push(argumentEnquote("releaseNotesFile", releaseNotesFile));
         }
 
-        configure.push(includeArguments(additionalArguments));
+        configure.push(includeAdditionalArgumentsAndProxyConfig(connection.url, additionalArguments));
 
         const code: Number = await octo
             .map((x) => x.launchOcto(configure))

--- a/source/tasks/Deploy/DeployV3/index.ts
+++ b/source/tasks/Deploy/DeployV3/index.ts
@@ -1,6 +1,6 @@
 import * as tasks from "azure-pipelines-task-lib/task";
 import * as utils from "../../Utils";
-import { multiArgument, connectionArguments, includeArguments, flag, argumentEnquote, argumentIfSet } from "../../Utils/tool";
+import { multiArgument, connectionArguments, includeAdditionalArgumentsAndProxyConfig, flag, argumentEnquote, argumentIfSet } from "../../Utils/tool";
 
 async function run() {
     try {
@@ -26,7 +26,7 @@ async function run() {
             multiArgument(argumentEnquote, "tenant", deploymentForTenants),
             multiArgument(argumentEnquote, "tenanttag", deployForTenantTags),
             flag("progress", showProgress),
-            includeArguments(additionalArguments),
+            includeAdditionalArgumentsAndProxyConfig(connection.url, additionalArguments),
         ];
 
         const code: Number = await octo

--- a/source/tasks/Deploy/DeployV4/index.ts
+++ b/source/tasks/Deploy/DeployV4/index.ts
@@ -1,6 +1,6 @@
 import * as tasks from "azure-pipelines-task-lib/task";
 import * as utils from "../../Utils";
-import { multiArgument, connectionArguments, includeArguments, flag, argumentEnquote, argumentIfSet } from "../../Utils";
+import { multiArgument, connectionArguments, includeAdditionalArgumentsAndProxyConfig, flag, argumentEnquote, argumentIfSet } from "../../Utils";
 
 async function run() {
     try {
@@ -14,7 +14,6 @@ async function run() {
         const deployForTenantTags = utils.getOptionalCsvInput("DeployForTenantTags");
         const showProgress = tasks.getBoolInput("ShowProgress");
         const additionalArguments = tasks.getInput("AdditionalArguments");
-
         await utils.assertOctoVersionAcceptsIds();
         const octo = await utils.getOrInstallOctoCommandRunner("deploy-release");
 
@@ -27,7 +26,7 @@ async function run() {
             multiArgument(argumentEnquote, "tenant", deployForTenants),
             multiArgument(argumentEnquote, "tenanttag", deployForTenantTags),
             flag("progress", showProgress),
-            includeArguments(additionalArguments),
+            includeAdditionalArgumentsAndProxyConfig(connection.url, additionalArguments),
         ];
 
         const code: Number = await octo

--- a/source/tasks/OctoCli/index.ts
+++ b/source/tasks/OctoCli/index.ts
@@ -1,6 +1,6 @@
 import * as tasks from "azure-pipelines-task-lib/task";
 import * as utils from "../Utils";
-import { connectionArguments, includeArguments } from "../Utils/tool";
+import { connectionArguments, includeAdditionalArguments } from "../Utils/tool";
 
 async function run() {
     try {
@@ -9,7 +9,7 @@ async function run() {
         const command = tasks.getInput("command", true);
         const octo = await utils.getOrInstallOctoCommandRunner(command);
 
-        const configure = [connectionArguments(connection), includeArguments(args)];
+        const configure = [connectionArguments(connection), includeAdditionalArguments(args)];
 
         const code: Number = await octo
             .map((x) => x.launchOcto(configure))

--- a/source/tasks/Pack/index.ts
+++ b/source/tasks/Pack/index.ts
@@ -1,9 +1,8 @@
 import * as tasks from "azure-pipelines-task-lib/task";
 import * as fs from "fs";
 import * as utils from "../Utils";
-import { argument, argumentIfSet, flag, multiArgument, argumentEnquote } from "../Utils";
+import { argument, argumentIfSet, flag, multiArgument, argumentEnquote, includeAdditionalArguments } from "../Utils";
 import { ToolRunner } from "azure-pipelines-task-lib/toolrunner";
-import { includeArguments } from "../Utils";
 
 export interface PackageRequiredInputs {
     packageId: string;
@@ -40,7 +39,7 @@ export const configure = (inputs: PackageInputs) => {
         argumentIfSet(argumentEnquote, "description", inputs.nuGetDescription),
         argumentIfSet(argumentEnquote, "releaseNotes", inputs.nuGetReleaseNotes),
         argument("overwrite", inputs.overwrite.toString()),
-        includeArguments(inputs.additionalArguments),
+        includeAdditionalArguments(inputs.additionalArguments),
         (tool: ToolRunner) => {
             if (!utils.isNullOrWhitespace(inputs.nuGetReleaseNotesFile) && fs.existsSync(inputs.nuGetReleaseNotesFile) && fs.lstatSync(inputs.nuGetReleaseNotesFile).isFile()) {
                 console.log(`Release notes file: ${inputs.nuGetReleaseNotesFile}`);

--- a/source/tasks/Promote/PromoteV3/index.ts
+++ b/source/tasks/Promote/PromoteV3/index.ts
@@ -1,6 +1,6 @@
 import * as tasks from "azure-pipelines-task-lib/task";
 import * as utils from "../../Utils";
-import { multiArgument, connectionArguments, includeArguments, flag, argumentEnquote, argumentIfSet } from "../../Utils/tool";
+import { multiArgument, connectionArguments, includeAdditionalArgumentsAndProxyConfig, flag, argumentEnquote, argumentIfSet } from "../../Utils/tool";
 
 async function run() {
     try {
@@ -27,7 +27,7 @@ async function run() {
             multiArgument(argumentEnquote, "tenant", deploymentForTenants),
             multiArgument(argumentEnquote, "tenanttag", deployForTenantTags),
             flag("progress", showProgress),
-            includeArguments(additionalArguments),
+            includeAdditionalArgumentsAndProxyConfig(connection.url, additionalArguments),
         ];
 
         const code: Number = await octo

--- a/source/tasks/Promote/PromoteV4/index.ts
+++ b/source/tasks/Promote/PromoteV4/index.ts
@@ -1,6 +1,6 @@
 import * as tasks from "azure-pipelines-task-lib/task";
 import * as utils from "../../Utils";
-import { multiArgument, connectionArguments, includeArguments, flag, argumentEnquote, argumentIfSet } from "../../Utils";
+import { multiArgument, connectionArguments, includeAdditionalArgumentsAndProxyConfig, flag, argumentEnquote, argumentIfSet } from "../../Utils";
 
 async function run() {
     try {
@@ -27,7 +27,7 @@ async function run() {
             multiArgument(argumentEnquote, "tenant", deployForTenants),
             multiArgument(argumentEnquote, "tenanttag", deployForTenantTags),
             flag("progress", showProgress),
-            includeArguments(additionalArguments),
+            includeAdditionalArgumentsAndProxyConfig(connection.url, additionalArguments),
         ];
 
         const code: Number = await octo

--- a/source/tasks/Push/PushV3/index.ts
+++ b/source/tasks/Push/PushV3/index.ts
@@ -1,7 +1,7 @@
 import * as tasks from "azure-pipelines-task-lib/task";
 import * as utils from "../../Utils";
 
-import { multiArgument, connectionArguments, includeArguments, flag, argumentEnquote, argumentIfSet } from "../../Utils/tool";
+import { multiArgument, connectionArguments, includeAdditionalArgumentsAndProxyConfig, flag, argumentEnquote, argumentIfSet } from "../../Utils/tool";
 
 async function run() {
     try {
@@ -15,7 +15,7 @@ async function run() {
         const octo = await utils.getOrInstallOctoCommandRunner("push");
         const matchedPackages = await utils.resolveGlobs(packages);
 
-        const configure = [connectionArguments(connection), argumentIfSet(argumentEnquote, "space", space), multiArgument(argumentEnquote, "package", matchedPackages), flag("replace-existing", replace), includeArguments(additionalArguments)];
+        const configure = [connectionArguments(connection), argumentIfSet(argumentEnquote, "space", space), multiArgument(argumentEnquote, "package", matchedPackages), flag("replace-existing", replace), includeAdditionalArgumentsAndProxyConfig(connection.url, additionalArguments)];
 
         const code: Number = await octo
             .map((x) => x.launchOcto(configure))

--- a/source/tasks/Push/PushV4/index.ts
+++ b/source/tasks/Push/PushV4/index.ts
@@ -1,7 +1,7 @@
 import * as tasks from "azure-pipelines-task-lib/task";
 import * as utils from "../../Utils";
 
-import { multiArgument, connectionArguments, includeArguments, argument, argumentEnquote, argumentIfSet, getOverwriteModeFromReplaceInput } from "../../Utils";
+import { multiArgument, connectionArguments, includeAdditionalArgumentsAndProxyConfig, argument, argumentEnquote, argumentIfSet, getOverwriteModeFromReplaceInput } from "../../Utils";
 
 async function run() {
     try {
@@ -16,7 +16,7 @@ async function run() {
         const octo = await utils.getOrInstallOctoCommandRunner("push");
         const matchedPackages = await utils.resolveGlobs(packages);
 
-        const configure = [connectionArguments(connection), argumentIfSet(argumentEnquote, "space", space), multiArgument(argumentEnquote, "package", matchedPackages), argument("overwrite-mode", overwriteMode), includeArguments(additionalArguments)];
+        const configure = [connectionArguments(connection), argumentIfSet(argumentEnquote, "space", space), multiArgument(argumentEnquote, "package", matchedPackages), argument("overwrite-mode", overwriteMode), includeAdditionalArgumentsAndProxyConfig(connection.url, additionalArguments)];
 
         const code: Number = await octo
             .map((x) => x.launchOcto(configure))

--- a/source/tasks/Utils/install.ts
+++ b/source/tasks/Utils/install.ts
@@ -5,6 +5,7 @@ import * as tools from "azure-pipelines-tool-lib/tool";
 import * as TypedRestClient from "typed-rest-client/RestClient";
 import { ToolName, ToolNameBeforeV7 } from "../Utils";
 import { head, filter } from "ramda";
+import { IProxyConfiguration } from "typed-rest-client/Interfaces";
 
 interface LatestResponse {
     latest: string;
@@ -28,8 +29,6 @@ interface Dictionary {
 
 const RestClient = TypedRestClient.RestClient;
 const OctopurlsUrl = "https://g.octopushq.com";
-
-var octopurls = new RestClient("OctoTFS/Tasks", OctopurlsUrl);
 
 const applyTemplate = (dictionary: Dictionary, template: string) => {
     return Object.keys(dictionary).reduce((result, key) => result.replace(new RegExp(`{\s*${key}\s*}`, "g"), dictionary[key] ? String(dictionary[key]) : ""), template);
@@ -96,13 +95,27 @@ async function resolvePublishedOctoVersion(version?: string): Promise<DownloadOp
     }
     console.log(`Attempting to contact ${OctopurlsUrl} to find Octopus CLI tool version ${version}`);
 
+    const proxyConfiguration = tasks.getHttpProxyConfiguration(OctopurlsUrl);
+    let proxySettings: IProxyConfiguration | undefined = undefined;
+
+    if(proxyConfiguration) {
+        proxySettings = {
+            proxyUrl: proxyConfiguration.proxyUrl,
+            proxyUsername: proxyConfiguration.proxyUsername,
+            proxyPassword: proxyConfiguration.proxyPassword,
+            proxyBypassHosts: proxyConfiguration.proxyBypassHosts,
+        }
+    }
+
+    const octopurls = new RestClient("OctoTFS/Tasks", OctopurlsUrl, undefined, { proxy: proxySettings });
+
     const response = await octopurls.get<LatestResponse>("LatestTools");
 
     if (response.result === null || response.result === undefined) {
         throw Error(`Failed to resolve Octopus CLI tool version ${version}. Endpoint returned status code ${response.statusCode})`);
     }
 
-    var option = filterPortableDownload(response.result.downloads);
+    const option = filterPortableDownload(response.result.downloads);
 
     if (option === null || option === undefined) {
         throw Error(`Failed to resolve the Octopus CLI tool portable download location. The result did not contain the download location.`);
@@ -113,7 +126,7 @@ async function resolvePublishedOctoVersion(version?: string): Promise<DownloadOp
     }
 
     //Adjust the version and location to point to the specified version
-    let result = { ...option, version };
+    const result = { ...option, version };
     result.location = applyTemplate(result, result.template);
 
     return result;

--- a/source/tasks/Utils/tool.ts
+++ b/source/tasks/Utils/tool.ts
@@ -128,7 +128,27 @@ export const argumentEnquote = curry((name: string, value: string | null | undef
     return argument(name, `"${value}"`, tool);
 });
 
-export const includeArguments = curry((value: string, tool: ToolRunner) => {
+export const includeAdditionalArgumentsAndProxyConfig = curry((url: string, value: string, tool: ToolRunner) => {
+    const proxyRegex = /-proxy=/;
+
+    const proxyConfig = tasks.getHttpProxyConfiguration(url)
+
+    if(proxyConfig) {
+        if (!proxyRegex.test(value)){
+            argument("proxy", proxyConfig.proxyUrl, tool);
+            if(proxyConfig.proxyUsername) {
+                argument("proxyUser", proxyConfig.proxyUsername, tool);
+            }
+            if(proxyConfig.proxyPassword){
+                argument("proxyPass", proxyConfig.proxyPassword, tool);
+            }
+        }
+    }
+
+    return includeAdditionalArguments(value, tool);
+});
+
+export const includeAdditionalArguments = curry((value: string, tool: ToolRunner) => {
     return tool.line(value);
 });
 

